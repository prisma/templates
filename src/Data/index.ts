export * as Data from './index_'
