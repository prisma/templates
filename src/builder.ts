import { BaseTemplateParametersResolved, BuildTemplate, File, Template } from './types'
import { Index, mapValues } from './utils'

/**
 * Build the given template with given parameters.
 */
export const build = <T extends Template>(
  template: T,
  parameters: GlobalPrismaTemplateParameters[T['metadata']['name']]
): BuildTemplate<T> => {
  const parameters_ = {
    ...parameters,
    ...template.Parameters.defaults,
  }

  return {
    ...template,
    files: runFileTransformers(transformers, template.files, parameters_),
  }
}

// Template Transformers

const prismaSchemaProviderTransformer: FileTransformer = (params) => {
  const { file, parameters } = params

  let content = file.content

  if (file.path === 'prisma/schema.prisma') {
    content = content.replace(`provider = "postgresql"`, `provider = "${parameters.datasourceProvider}"`)

    if (parameters.datasourceProvider === 'sqlserver') {
      content = content.replace(
        `provider = "prisma-client-js"`,
        `provider = "prisma-client-js"
  previewFeatures = ["microsoftSqlServer"]`
      )
    }
  }

  return {
    ...file,
    content,
  }
}

const transformers = [prismaSchemaProviderTransformer]

// Template Transformers System

type TransformerParams = { file: File; parameters: BaseTemplateParametersResolved }

type FileTransformer = (params: TransformerParams) => File

const runFileTransformers = (
  transformers: FileTransformer[],
  files: Index<File>,
  parameters: TransformerParams['parameters']
) => {
  return mapValues(files, (file) => {
    return {
      ...file,
      content: transformers.reduce((file, transformer): File => {
        return transformer({
          file,
          parameters,
        })
      }, file),
    }
  })
}
