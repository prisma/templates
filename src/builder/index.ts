export * as Builder from './builder'
