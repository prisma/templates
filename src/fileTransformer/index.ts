export * as FileTransformer from './fileTransformer'
