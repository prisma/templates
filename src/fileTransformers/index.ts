export * as FileTransformers from './index_'
