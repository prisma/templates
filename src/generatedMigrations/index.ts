export * as MigrationsSql from './index_'