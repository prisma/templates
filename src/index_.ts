import { Generated } from './generated'

export * as $Types from './types'

export const Templates = Generated.Templates

export const Utils = {
  getVercelDeployButtonUrl(params: { repositoryOwner: string; repositoryHandle: string }) {
    const repository = encodeURIComponent(params.repositoryOwner + '/' + params.repositoryHandle)
    const envVars = ['DATABASE_URL', 'DATABASE_MIGRATE_URL'].join(',')
    const envVarDescription = encodeURIComponent(
      "Connection string for the database this deployment will talk to. If you're using the Prisma Data Proxy, then use its connection string."
    )

    return `https://vercel.com/new/import?repository-url=https%3A%2F%2Fgithub.com%2F${repository}&env=${envVars}&envDescription=${envVarDescription}`
  },
}
