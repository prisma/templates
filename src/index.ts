export * as PrismaTemplates from './index_'
import * as PrismaTemplates from './index_'

const x = new PrismaTemplates.Templates.Empty({})
