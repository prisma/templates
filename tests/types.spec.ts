import { PrismaTemplates } from '~/src'

it('has handle data lists', () => {
  expect(PrismaTemplates.$Types.TemplateHandle).toMatchSnapshot()
})
