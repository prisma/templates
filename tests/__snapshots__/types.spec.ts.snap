// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`has handle data lists 1`] = `
Object {
  "Camel": Array [
    "nextjs",
    "empty",
    "musicStreamingService",
    "rentalsPlatform",
    "saas",
    "urlShortener",
  ],
  "Kebab": Array [
    "nextjs",
    "empty",
    "music-streaming-service",
    "rentals-platform",
    "saas",
    "url-shortener",
  ],
  "Pascal": Array [
    "Nextjs",
    "Empty",
    "MusicStreamingService",
    "RentalsPlatform",
    "Saas",
    "UrlShortener",
  ],
  "Snake": Array [
    "nextjs",
    "empty",
    "music_streaming_service",
    "rentals_platform",
    "saas",
    "url_shortener",
  ],
  "Upper": Array [
    "NEXTJS",
    "EMPTY",
    "MUSIC_STREAMING_SERVICE",
    "RENTALS_PLATFORM",
    "SAAS",
    "URL_SHORTENER",
  ],
}
`;
